packages/lit-dev-content/site/tutorials/content/reactivity/04.md

```diff
@@ -0,0 +1,37 @@
+One way your component can react to property changes is to compute values based on the new and existing property values. 
+
+The `willUpdate()` callback is the ideal place to do these calculations in Lit. It's called near the beginning up the update cycle, before the component renders, so any properties you calculate during `willUpdate()` are available in your `render()` method.
+
+The `willUpdate()` method receives a map of changed properties as an argument, where the keys are property names and the values are the old values for each changed property. In this case, we'll compute a hash value based on the `name` property.
+
+First, import the `computeSHA()` function:
+
+```js
+import {computeSHA} from './compute-sha.js';
+```
+
+Then add the `willUpdate()` callback:
+
+{% switchable-sample %}
+
+```ts
+  protected willUpdate(changedProperties: PropertyValues<this>): void {
+    // Only need to check changed properties for an expensive computation.
+    if (changedProperties.has('name')) {
+      computeSHA(this.name).then((sha) => this.sha = sha);
+    }
+  }
```

Comment:
Yeah... I like the idea of a synchronous action much better. I was having trouble justifying, say, rot13ing the name, because why not just do that in `render()`? But I think it would make a simpler demo without the weird note about async actions...

The example in the docs (which I think is from @sorvell but I could be wrong) uses a `computeSHA` function so that's what I tried to kludge together. I dunno if @sorvell has thoughts on this.

Reply:
Makes sense. I think I'll use Elliott's forward/backward text example... It's a little funky, but we've done first name/last name/name tag so many times in the docs that I think it's nice to take a break.

(If syncing two ways is too complicated, I could always just make it rot-13.)

Reply:
Went with @e111077 's demo, it's kind of random and fun.